### ISSUE TITLE
add coroutines support

### ISSUE THREAD

**Issue by pohmelie** [cool-RR/PySnooper#27]
Coroutines are first class citizens in python world.

**Comment by cool-RR** [comment]
Sounds good, send a pull request.

**Comment by PetterS** [comment]
The code for this will be cleaner once only Python 3.5+ is supported.

**Comment by cool-RR** [comment]
Yes, unfortunately a big use case for PySnooper is shitty legacy projects, so keeping Python 2.7 support is important.

**Comment by pohmelie** [comment]
I have successfully patched pysnooper to support coroutines, looks neat, but requires 3.5+
``` python
def snoop(output=None, variables=(), depth=1, prefix=''):

    @contextlib.contextmanager
    def _tracer_context(function):
        write = get_write_function(output)
        target_code_object = function.__code__
        with Tracer(target_code_object=target_code_object,
                    write=write, variables=variables,
                    depth=depth, prefix=prefix):
            yield

    def decorator(function):

        @functools.wraps(function)
        async def awrapper(*args, **kwargs):
            with _tracer_context(function):
                return await function(*args, **kwargs)

        @functools.wraps(function)
        def wrapper(*args, **kwargs):
            with _tracer_context(function):
                return function(*args, **kwargs)

        if inspect.iscoroutinefunction(function):
            return awrapper
        else:
            return wrapper

    return decorator
```

**Comment by cool-RR** [comment]
Great work.

1. It's possible to make this not break Python 2 by executing code with
`exec`. I can do that so you can leave that to the end.
2. I want the tracer context manager to be acquired only when the coroutine
is active.
3. If you'd like to implement generator support on the way, that'd be nice.
Not mandatory.
4. I want tests for everything. (Again, feel free to ignore Python 2
support for now, I'll deal with that later.)

**Comment by pohmelie** [comment]
Ok, I will try.

**Comment by Martin9527** [comment]
so,what about now?do we support coroutines in python 2.7?

**Comment by cool-RR** [comment]
@Martin9527 There are no native coroutines in Python 2.7, just generators which can be used as coroutines.

**Comment by Martin9527** [comment]
> @Martin9527 There are no native coroutines in Python 2.7, just generators which can be used as coroutines.

yeah，we defined coroutines based on generators，so can this tools support for generators？

**Comment by cool-RR** [comment]
> so can this tools support for generators？

Right now it doesn't support generators, but I'd welcome a pull request that'd support generators, or preferably combine it in this pull request.

**Comment by cool-RR** [comment]
Whoever is interested in doing this issue, this might help: https://stackoverflow.com/questions/56078482/python-coroutines-release-context-manager-when-pausing/

**Comment by Chise1** [comment]
When will it could be finished?

**Comment by cool-RR** [comment]
> When will it could be finished?

Either tomorrow or never, depending on you.
